R/aggregate.R

```diff
@@ -180,43 +143,380 @@ agg_climate_data <- function(year, data_raster, climate_var, daily_agg, trans =
+
+
+  # Define spline function
+  get_spline <- function(x){
+
+    # Make first raster returned just the climate variable to preserve it's column in the resulting data table
+    if(x == 0){
+      return(clim_daily)
+    }
+    # Add in spline terms, all of which are 0 if negative
+    else{
+      clim_daily_table <- raster::values(clim_daily)
+
+      clim_daily_table <-ifelse((clim_daily_table - knot_locs[x])^3 > 0,
+                                (clim_daily_table - knot_locs[x])^3, 0)
+      - ((ifelse((clim_daily_table - knot_locs[num_knots - 1])^3 > 0,
+                 (clim_daily_table - knot_locs[num_knots - 1])^3, 0))
+         *((knot_locs[num_knots] - knot_locs[x]))
+           / (knot_locs[num_knots] - knot_locs[num_knots - 1]))
+      + ((ifelse((clim_daily_table - knot_locs[num_knots])^3 > 0,
+                 (clim_daily_table - knot_locs[num_knots])^3, 0))
+         *((knot_locs[num_knots - 1] - knot_locs[x])
+           / (knot_locs[num_knots] - knot_locs[num_knots - 1])))
+
+      clim_daily_new <- clim_daily
+      raster::values(clim_daily_new) <- clim_daily_table
+
+      return(clim_daily_new)
+
+
+
+    }
+  }
+
+
+  # For each layer, create new spline variables
+  r <- lapply(0:list_length, get_spline)
+
+
+  ## Function: Set names of data table by month, change from wide to long format, rename based on polynomial orders
+  create_dt <- function(x){
+
+    # Should output raster cells x/y with 365 days as column names
+    dt <- as.data.table.raster(r[[x]], xy=TRUE)
+
+    # Set column names with months
+    new_names <- c('x', 'y', layer_names)
+    data.table::setnames(dt, new_names)
+
+    # Change from wide to long format
+    dt = data.table::melt(dt, id.vars = c("x", "y"))
+
+    # Update variable names
+    var_names <- c('date', list_names[x])
+    data.table::setnames(dt, old=c('variable', 'value'), new=var_names)
+  }
+
+  # Make each raster layer a data.table
+  list_dt <- lapply(1:(list_length + 1), create_dt)
+
+  # Merge all data tables together
+  clim_dt <- list_dt[[1]]
+
+  i <- 2
+  while(i <= list_length + 1){
+    dt_m <- list_dt[[i]]
+    clim_dt <- merge(clim_dt, dt_m, by=c('x', 'y', 'date'))
+    i <- i + 1
+  }
+
+
+
+  # Aggregate by polygon
+  sum_by_poly <- after_trans(clim_dt, weights_dt, list_names, second_weights, agg_to)
+
+  return(sum_by_poly)
+}
+
+
+
+#' @export
+staggregate_bin <- function(data_raster, climate_var, daily_agg, num_bins = 30, binwidth = NULL, center_on = NULL, start_on = NULL, end_on = NULL, max = NULL, min = NULL, geoweights_table, second_weights, agg_to = "month"){
+  # Get climate data as a data table and aggregate to daily values
+  setup_list <- before_trans(data_raster, climate_var, daily_agg, geoweights_table, second_weights)
+  clim_daily <- setup_list[[1]]
+  weights_dt <- setup_list[[2]]
+  layer_names <- setup_list[[3]]
+
+
+  clim_daily_table <- raster::values(clim_daily)
```

Comment:
like above, @tliddell4 flagging i will read this transformation in detail but cannot right now.